front/src/utils/types.ts

```diff
@@ -15,3 +17,33 @@ export type ArrayElement<ArrayType extends readonly unknown[] | undefined | null
 export type typedEntries<T> = Array<[keyof T, T[keyof T]]>;
 
 export type Unit = 'meters' | 'millimeters';
+
+export const deepCloneMap = <K, T>(map: Map<K, T>) => {
+  const clonedMap = new Map<K, T>();
+  for (const [key, value] of map) {
+    if (value instanceof Object) {
+      clonedMap.set(key, JSON.parse(JSON.stringify(value)));
+    } else {
+      clonedMap.set(key, value);
+    }
+  }
+  return clonedMap;
+};
+
+export const mapBy = <K extends keyof T, T>(items: T[] | undefined, key: K): Map<T[K], T> => {
+  const map = new Map<T[K], T>();
+  if (items) {
+    items.forEach((item) => {
+      map.set(item[key], item);
+    });
+  }
+  return map;
+};
+
+export const concatMap = <K, V>(map1: Map<K, V>, map2: Map<K, V>) => {
+  const result = new Map<K, V>(map1);
+  for (const [key, value] of map2) {
+    result.set(key, value);
+  }
+  return result;
+};
```

Comment:
A simpler way to write this would be:

```ts
new Map([...map1.entries(), ...map2.entries()])
```

at which point it's probably trivial enough to be inlined at the call-site.

Reply:
Thanks :+1: